### ISSUE TITLE
TTS example

### ISSUE THREAD

**Issue by Purus** [JonnyBurger/remotion#249]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%24115%20Rewarded-%237E24E3.svg" />](https://issuehunt.io/r/JonnyBurger/remotion/issues/249)

Would be great if there is support for converting text to speech automatically and include as part of the video output.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $115.00)

- [<img src='https://avatars.githubusercontent.com/u/1629785?v=4' alt='jonnyburger' width=24 height=24> jonnyburger](https://issuehunt.io/u/jonnyburger) ($115.00)

### Submitted pull Requests
- [#427 Bogus PR for https://github.com/JonnyBurger/remotion/issues/249](https://issuehunt.io/r/JonnyBurger/remotion/pull/427)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/JonnyBurger/remotion/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by JonnyBurger** [comment]
We discussed it a bit here:

https://github.com/JonnyBurger/remotion/issues/28#issuecomment-813941193

I found that there is non suitable browser API but of course you can use a cloud service from Google and AWS. I'm gonna leave this open because it's a great reminder to build a starter template for it and release it 🙂

**Comment by Purus** [comment]
Would be great if we could not be dependent on third-party cloud services.. May be something like https://github.com/MikeyParton/react-speech-kit

**Comment by JonnyBurger** [comment]
Here we are constrained by the APIs of the browser.
While you can trigger TTS, you cannot access the audio data, put it in an audio tag, export it as a file, etc. :( This makes it impossible to render in a video.

**Comment by tohagan** [comment]
To do TTS well requires that it can coordinate with animation timing.  You’re likely to eventually combine automated translation with TSS and that means the audio lengths will vary in time so you need a means to time associated animation to the length of the resulting audio. You may also wish to display or highlight specific words (which may be translated) in sync with the matching audio.

**Comment by tohagan** [comment]
If you’re combining word animation with translation, you want to perform the translation in one API call if possible, not just for performance but to provide the best translation context so, for example, the sequence of related words being animated are translated together.

**Comment by JonnyBurger** [comment]
I am adding a $60 bounty to whoever submits a text-to-speech example that works in preview mode as well as render mode and that uses a cloud TTS API.
(We should be able to release the example as part of Remotion but of course you will maintain credit)

**Comment by tohagan** [comment]
To get the timing of the words spoken, you can reverse the process and use STT APIs (Google supports 125 languages and accents). Their API gives you a list of phrases spoken (“alternatives” array) and then within each phrase the timing of each individual word. So you can use these as the basis of the text animation timing.  Of course STT AI is less reliable than TTS so the words may not always match the original input text sent to TTS so some manual or smart match up editing may be required.  I’ve just used this timing info in a prototype app I’m working on for videos last Friday.

**Comment by FelippeChemello** [comment]
@JonnyBurger What do you think about credentials for this TTS services? Should store in .env settings and, when running video and use TTS check if it is set?

**Comment by JonnyBurger** [comment]
Good question!

We don't currently have support for .env files (but this is a great idea, I will create another issue for this). So the best way for the moment I think is using input props: https://www.remotion.dev/docs/parametrized-rendering#input-props

**Comment by FelippeChemello** [comment]
Do you see any problem at using 2 different voices? One at preview (SpeechSyntesis) and another during render (Cloud Voice).
I was thinking about it and found some problem at this, because for render I need to download file, however I can't during preview since I haven't access to filesystem

**Comment by tohagan** [comment]
Cloud Voice audio output can be saved to the cloud file system as a cache audio file or downloaded locally. Audio file name could be a hash of input text and voice params and so only recomputed when text/params change. You can then stream the local or cloud audio file.

**Comment by tohagan** [comment]
## Swapping Chrome APIs with Cloud APIs

One thing to keep in mind as you plan the roadmap for TTS (and perhaps STT) features is that there are important differences between what you can achieve with [SpeechSynthesis](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesis) and a corresponding cloud service. For simple TTS they could be swapped.  Caching audio files generated from cloud service could improve the DX which may be sufficient for a basic MVP solution. However I recommend that you look further down the roadmap to consider where these cloud services might take you as you design the current Remotion speech architecture and APIs.

## Cloud service Audio Events

Cloud TTS services not only generate audio but they can also be used to generate associated events timed to correlated with the audio. Also different cloud services deliver different features.

### Word/Phrase Events
Using STT you can emit a stream of words and phrases with associated audio track timing. This is typically used to generate subtitles but can be repurposed for animation. This could be used in conjunction with TTS or independently on any audio source containing speech. By pre-capturing these timed events, a frame render might wish to compute the time **until** the next word/phrase occurs or **since** the last word/phrase occurred and then use this to animate word text or other related object animations. This may be particularly important when handling text or audio translations as the speech timing can vary widely between translations.  

### Lip Sync Events
Azure STT can emit a [stream of viseme lip sync events](https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/how-to-speech-synthesis-viseme?pivots=programming-language-javascript) that can be used to animate lips on a 2D or 3D avatar. 

### Audio / Video track events
Many existing audio & video files store timed events that could be be very useful in animation. 

In summary, rendering may need to supporting multi-tracking timed events as input for the rendering process with the ability to compute time to previous and next events and associated event metadata.

**Comment by FelippeChemello** [comment]
I had working on it. Here is the example https://github.com/FelippeChemello/Remotion-TTS-Example. 
At IssueHunt asks me for making a PR, however I think that it wasn't the purpose of it Issue. How can I proceed @JonnyBurger ?

**Comment by JonnyBurger** [comment]
@FelippeChemello Wow thank you! This looks super awesome!
I can only pay out if you submit a PR apparently, can you submit a bogus PR?

I will close it but pay it out immediately 🤝 

Plus, it would be cool, if I could become a collaborator of the repository, so that should it be necessary, I can adjust the code, upgrade dependencies or change the README.
Our goal is to provide a streamlined set of template for different usecases.
